tiledb/sm/query/sparse_global_order_reader.cc

```diff
@@ -0,0 +1,918 @@
+  // No subarray means we process all cells.
+  if (!subarray_set)
+    return Status::Ok();
+
+  // For easy reference.
+  auto coords_num = tile->cell_num();
+  auto dim_num = array_schema_->dim_num();
+  auto cell_order = array_schema_->cell_order();
+  auto range_coords = subarray_.get_range_coords(0);
+
+  std::vector<uint8_t> result_bitmap(coords_num, 1);
+
+  // Compute result and overwritten bitmap per dimension
+  for (unsigned d = 0; d < dim_num; ++d) {
+    // For col-major cell ordering, iterate the dimensions
+    // in reverse.
+    const unsigned dim_idx =
+        cell_order == Layout::COL_MAJOR ? dim_num - d - 1 : d;
+    const auto& ranges = subarray_.ranges_for_dim(dim_idx);
+    RETURN_NOT_OK(tile->compute_results_sparse(
+        dim_idx, ranges[range_coords[dim_idx]], &result_bitmap, cell_order));
+  }
+
+  tiles_result_bitmap[tile->frag_idx()] = std::move(result_bitmap);
+
+  return Status::Ok();
+}
+
+template <class T>
+Status SparseGlobalOrderReader::compute_result_cell_slabs(
+    uint64_t memory_budget, T cmp) {
+  auto allows_dups = array_schema_->allows_dups();
+
+  // For easy reference.
+  bool subarray_set = subarray_.is_set();
+
+  // A tile min heap, contains one Result coords per fragment.
+  std::priority_queue<ResultCoords, std::vector<ResultCoords>, T> tile_queue(
+      cmp);
+
+  // Tile iterators, per fragments.
+  std::vector<std::list<ResultTile>::iterator> result_tiles_it(
+      result_tiles_.size());
+
+  // Bitmaps for the current tile, per fragment for the cells to process.
+  std::vector<std::vector<uint8_t>> tiles_result_bitmap(result_tiles_.size());
+
+  // For all fragments, get the first tile.
+  for (unsigned int frag_idx = 0; frag_idx < result_tiles_.size(); frag_idx++) {
+    if (result_tiles_[frag_idx].size() > 0) {
+      // Initialize the iterator for this fragment.
+      result_tiles_it[frag_idx] = result_tiles_[frag_idx].begin();
+      auto tile = &*result_tiles_it[frag_idx];
+
+      // Get the cell index we were processing.
+      auto cell_idx = frag_tile_idx_[frag_idx].second;
+
+      // Calculate the bitmap for the cells.
+      auto rc = ResultCoords(tile, cell_idx);
+      RETURN_NOT_OK(
+          compute_tile_result_bitmap(subarray_set, tile, tiles_result_bitmap));
+
+      // Calculate hilbert values, this is templated out for non hilbert code.
+      RETURN_NOT_OK(calculate_hilbert_values(
+          subarray_set, tile, tiles_result_bitmap, cmp));
+
+      // Add the tile to the queue.
+      tile_queue.emplace(std::move(rc));
+    }
+  }
+
+  // The result cell slab in progress.
+  ResultCellSlab rcs =
+      ResultCellSlab(tile_queue.top().tile_, tile_queue.top().pos_, 0);
+
+  // Used to store cells that have the same coordinate as the one in process.
+  std::stack<ResultCoords> same_coords_stack;
+
+  // Process all elements.
+  bool need_more_tiles = false;
+  while (!tile_queue.empty() && !need_more_tiles) {
```

Comment:
I am a bit worried about performance. This is serial and we the number of cells (especially when we consolidate) will just be massive. I think we need a more sophisticated algorithm where we identify as many result cells as possible can fit in the result, and try to sort in parallel as we used to. I will provide more insights online.